src/apps/lwaftr/lwaftr.lua

```diff
@@ -134,6 +200,129 @@ local function get_icmp_payload(ptr)
    return ptr + constants.icmp_base_size
 end
 
+local function write_ethernet_header(pkt, ether_type)
+   local h = ffi.cast(ethernet_header_ptr_t, pkt.data)
+   ffi.fill(h.shost, 6, 0)
+   ffi.fill(h.dhost, 6, 0)
+   h.type = ether_type
+end
+
+local function prepend_ethernet_header(pkt, ether_type)
+   pkt = packet.shiftright(pkt, ethernet_header_size)
+   write_ethernet_header(pkt, ether_type)
+   return pkt
+end
+
+local function write_ipv6_header(ptr, src, dst, tc, next_header, payload_length)
+   local h = ffi.cast(ipv6_header_ptr_t, ptr)
+   h.v_tc_fl = 0
+   lib.bitfield(32, h, 'v_tc_fl', 0, 4, 6)   -- IPv6 Version
+   lib.bitfield(32, h, 'v_tc_fl', 4, 8, tc)  -- Traffic class
+   lib.bitfield(32, h, 'v_tc_fl', 12, 20, 0) -- Flow label
+   h.payload_length = htons(payload_length)
+   h.next_header = next_header
+   h.hop_limit = constants.default_ttl
+   h.src_ip = src
+   h.dst_ip = dst
+end
+
+local function calculate_icmp_payload_size(dst_pkt, initial_pkt, max_size, config)
+   local original_bytes_to_skip = ethernet_header_size
+   if config.extra_payload_offset then
+      original_bytes_to_skip = original_bytes_to_skip + config.extra_payload_offset
+   end
+   local payload_size = initial_pkt.length - original_bytes_to_skip
+   local non_payload_bytes = dst_pkt.length + constants.icmp_base_size
+   local full_pkt_size = payload_size + non_payload_bytes
+   if full_pkt_size > max_size + ethernet_header_size then
+      full_pkt_size = max_size + ethernet_header_size
+      payload_size = full_pkt_size - non_payload_bytes
+   end
+   return payload_size, original_bytes_to_skip, non_payload_bytes
+end
+
+-- Write ICMP data to the end of a packet
+-- Config must contain code and type
+-- Config may contain a 'next_hop_mtu' setting.
+
+local function write_icmp(dst_pkt, initial_pkt, max_size, base_checksum, config)
+   local payload_size, original_bytes_to_skip, non_payload_bytes =
+      calculate_icmp_payload_size(dst_pkt, initial_pkt, max_size, config)
+   local off = dst_pkt.length
+   dst_pkt.data[off] = config.type
```

Comment:
This could be rewritten using a icmp header instead of offset style. I understand this PR simply moves some modules to lwaftr.lua and this could be done in a next PR.

Reply:
Agreed!  I am working on this but the need to build ethernet headers everywhere was bothering me, and so I will retake this point after doing the refactoring discussed in #922 .